java/src/test/java/org/rocksdb/FlinkCompactionFilterTest.java

```diff
@@ -0,0 +1,344 @@
+import org.junit.rules.TemporaryFolder;
+import org.rocksdb.FlinkCompactionFilter.StateType;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class FlinkCompactionFilterTest {
+    private static final String MERGE_OPERATOR_NAME = "stringappendtest";
+    private static final byte DELIMITER = ',';
+    private static final long TTL = 100;
+
+    private List<StateContext> stateContexts;
+    private List<ColumnFamilyDescriptor> cfDescs;
+    private List<ColumnFamilyHandle> cfHandles;
+
+    @Rule
+    public TemporaryFolder dbFolder = new TemporaryFolder();
+
+    @Before
+    public void init() {
+        stateContexts = new ArrayList<>(StateType.values().length);
+        cfDescs = new ArrayList<>();
+        cfHandles = new ArrayList<>();
+        cfDescs.add(new ColumnFamilyDescriptor(RocksDB.DEFAULT_COLUMN_FAMILY));
+        for (StateType type : StateType.values()) {
+            if (type != StateType.Disabled) {
+                StateContext stateContext = StateContext.create(type);
+                stateContexts.add(stateContext);
+                cfDescs.add(stateContext.getCfDesc());
+            }
+        }
+    }
+
+    @After
+    public void cleanup() {
+        for (StateContext stateContext : stateContexts) {
+            stateContext.cfDesc.getOptions().close();
+            stateContext.filterFactory.close();
+        }
+    }
+
+    @Test
+    public void checkStateTypeEnumOrder() {
+        // if the order changes it also needs to be adjusted
+        // in utilities/flink/flink_compaction_filter.h
+        // and in utilities/flink/flink_compaction_filter_test.cc
+        assertThat(StateType.Disabled.ordinal()).isEqualTo(0);
+        assertThat(StateType.Value.ordinal()).isEqualTo(1);
+        assertThat(StateType.List.ordinal()).isEqualTo(2);
+    }
+
+    @Test
+    public void testCompactionFilter() throws RocksDBException {
+        try(DBOptions options = createDbOptions();
+            RocksDB rocksDb = setupDb(options)) {
+            try {
+                for (StateContext stateContext : stateContexts) {
+                    stateContext.updateValueWithTimestamp(rocksDb);
+                    stateContext.checkUnexpired(rocksDb);
+                    rocksDb.compactRange(stateContext.columnFamilyHandle);
+                    stateContext.checkUnexpired(rocksDb);
+                }
+
+                for (StateContext stateContext : stateContexts) {
+                    stateContext.expire();
+                    stateContext.checkUnexpired(rocksDb);
+                    rocksDb.compactRange(stateContext.columnFamilyHandle);
+                    stateContext.checkExpired(rocksDb);
+                    rocksDb.compactRange(stateContext.columnFamilyHandle);
+                }
+            } finally {
+                for (ColumnFamilyHandle cfHandle : cfHandles) {
```

Comment:
This finally block will run after the try-with-resource ended and the db was closed.